x/gov/handler.go

```diff
@@ -304,25 +373,132 @@ func (h TextProposalHandler) validate(ctx weave.Context, db weave.KVStore, tx we
 	return &msg, rules, elect, nil
 }
 
+type ElectorateUpdateProposalHandler struct {
+	auth        x.Authenticator
+	elecBucket  *ElectorateBucket
+	propBucket  *ProposalBucket
+	rulesBucket *ElectionRulesBucket
+}
+
+func (h ElectorateUpdateProposalHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	if _, _, _, err := h.validate(ctx, db, tx); err != nil {
+		return nil, err
+	}
+	return &weave.CheckResult{GasAllocated: proposalCost}, nil
+
+}
+
+func (h ElectorateUpdateProposalHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, rules, electorate, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+	blockTime, _ := weave.BlockTime(ctx)
+
+	props, err := h.propBucket.GetByElectorate(db, msg.ElectorateID)
+	if err != nil {
+		return nil, err
+	}
+	for _, v := range props {
+		if v.Status == Proposal_Submitted {
+			return nil, errors.Wrapf(errors.ErrInvalidState, "open proposal using this electorate: %q", v.Title)
+		}
+	}
+
+	proposal := &Proposal{
+		Type:            Proposal_UpdateElectorate,
+		Title:           msg.Title,
+		Description:     msg.Description,
+		ElectionRuleID:  electorate.UpdateElectionRuleID,
+		ElectorateID:    msg.ElectorateID,
+		VotingStartTime: msg.StartTime,
+		VotingEndTime:   msg.StartTime.Add(time.Duration(rules.VotingPeriodHours) * time.Hour),
+		SubmissionTime:  weave.AsUnixTime(blockTime),
+		Author:          msg.Author,
+		VoteState:       NewTallyResult(rules.Quorum, rules.Threshold, electorate.TotalElectorateWeight),
+		Status:          Proposal_Submitted,
+		Result:          Proposal_Undefined,
+		Details: &Proposal_ElectorateUpdateDetails{&ElectorateUpdatePayload{
+			msg.DiffElectors,
+		}},
+	}
+
+	obj, err := h.propBucket.Build(db, proposal)
+	if err != nil {
+		return nil, err
+	}
+	if err := h.propBucket.Save(db, obj); err != nil {
+		return nil, errors.Wrap(err, "failed to persist proposal")
+	}
+
+	res := &weave.DeliverResult{
+		Data: obj.Key(),
+		Tags: []common.KVPair{
+			{Key: []byte(tagProposalID), Value: obj.Key()},
+			{Key: []byte(tagProposer), Value: msg.Author},
```

Comment:
This is a nice custom tag, to search by author.
That said, if there is a secondary index with author, it is auto-tagged already.

(Don't say it is hard, this is all done with a bunch of helpers in iov-core already)

Reply:
Good to know. There is no index for author but I will add one.

Reply:
I have removed all manual tags and added an index for author 👍